pkg/test/framework/environments/kubernetes/flags.go

```diff
@@ -40,4 +40,10 @@ func init() {
 		"Deploy Istio into the target Kubernetes environment.")
 	flag.BoolVar(&globalSettings.MinikubeIngress, "istio.test.kube.minikubeingress", globalSettings.MinikubeIngress,
 		"Configure the Ingress component so that it gets the IP address from Node, when Minikube is used..")
+	flag.UintVar(&globalSettings.IngressGateway.ReplicaCount, "istio.test.kube.ingressGateway.replicaCount", globalSettings.IngressGateway.ReplicaCount,
```

Comment:
Adding all of these flags in this manner seems like a losing proposition. As we add more tests, we would end up needing more flags.

Can we keep this more generic instead? Some options to consider:

- Supply a values file instead.
- Allow passing in key==values 
- Allow test (specifically TestMain) to specify these values.

It looks like, based on the model we're gravitating towards, it might be better to allow passing some of these through TestMain. I think we can separate the Helm values broadly into two categories:

1. Values pertaining to the development environment (i.e. Hub Tag, ImagePullPolicy etc.). These needs to be variable, so that the tests can run in a multitude of environments reliably, but otherwise do not affect the semantics of the test.
2. Values pertaining to the things that are being tested. These should be a rigid set of settings per test. Given that we want to deploy Istio once per TestMain, it makes sense to declare these in TestMain.

If we expose last item through command-line flags, then I suspect we will end up in a place, where to run a test successfully, you'll need to specify the exact command-line flags.

WDYT?

Reply:
Agreed - I have been thinking something similar for a while now and I think we're finally getting to the point where we need to re-think this a bit.

Providing a values file sounds like a good approach in general. And IIRC helm will allow us to have a default file and then apply overrides. So we could allow the user to specify a values file on the command-line, which would be an override of the Istio defaults. This is probably a good start.

The other thing to consider is if a test requires a certain configuration (for example, this test is likely to fail unless the autoscale parameters match the replicaCount).  Should we be able to disable a test based on the values? In this particular case, we could rewrite the test so that the traffic is sent for a longer duration than our query range in order to force autoscaling up before the query range.  A test re-write may not always be possible, however. Perhaps these sorts of tests (i.e. conformance) are just kept separate from CI tests and are always run against a "known" environment.  I think that's probably fair for now.

Reply:
@ozevren would it be acceptable to you if we resolve this issue in a follow-on PR? I can create an issue to capture our thoughts.

Reply:
Yes, it is reasonable to track and resolve this issue independently, but let's make sure we have the right comments/TODOs/bugs in place to ensure that it is tracked properly, and people would not make assumptions about the correctness of those flags. I'll review the rest of it soon.